Sum Types  behave inconsistently in Closures

### Describe the bug

See repro code.

### Expected Behavior

```v
module main

import encoding.binary
import rand

type DataType = int | string

struct DataBuilder {
mut:
	func fn (mut []u8)
	len  int
}

pub fn (mut d DataBuilder) add(v DataType) {
	func := d.func
	len := d.len
	d.func = fn [v, len, func] (mut b []u8) {
		if !isnil(func) {
			func(mut b)
		}
		v_ := v
		match v_ {
			int {
				binary.big_endian_put_u32_at(mut b, u32(v_), len)
			}
			string {
				binary.big_endian_put_u16_at(mut b, u16(v_.len), len)
				for i := 0; i < v_.len; i++ {
					b[len + 2 + i] = v_[i]
				}
			}
		}
	}
	d.len += match v {
		int { 4 }
		string { 2 + v.len }
	}
}

pub fn (d &DataBuilder) build() []u8 {
	mut b := []u8{len: d.len}
	d.func(mut b)
	binary.big_endian_put_u16(mut b, u16(d.len - 2))
	unsafe {
		d.func = nil
		d.len = 2
	}
	return b
}

fn new_data_builder() DataBuilder {
	return DataBuilder{
		func: unsafe { nil }
		len: 2
	}
}

fn main() {
	mut data := new_data_builder()
	data.add(1)
	data.add(2)
	data.add(3)
	data.add(4)
	data.add(5)
	data.add('1')
	data.add('22')
	data.add('334455')
	data.add(rand.string(rand.int_in_range(10, 100)!))
	data.add('5')
	data.add(1)
	data.add(2)
	data.add(3)
	data.add(4)
	data.add(5)
	ret := data.build()
	dump(binary.big_endian_u16_at(ret, 0) == ret[2..].len)
	dump(ret)
}

```

### Current Behavior

```
==================
d:/tools/compiler/v/thirdparty/tcc/include/winapi/synchapi.h:167: warning: CreateMutex redefined
C:/Users/Administrator/AppData/Local/Temp/v_0/main.16890391755949642944.tmp.c:2542: error: 'v' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

### Reproduction Steps

```v
module main

import encoding.binary
import rand

type DataType = int | string

struct DataBuilder {
mut:
	func fn (mut []u8)
	len  int
}

pub fn (mut d DataBuilder) add(v DataType) {
	func := d.func
	len := d.len
	d.func = fn [v, len, func] (mut b []u8) {
		if !isnil(func) {
			func(mut b)
		}
		match v {
			int {
				binary.big_endian_put_u32_at(mut b, u32(v), len)
			}
			string {
				binary.big_endian_put_u16_at(mut b, u16(v.len), len)
				for i := 0; i < v.len; i++ {
					b[len + 2 + i] = v[i]
				}
			}
		}
	}
	d.len += match v {
		int { 4 }
		string { 2 + v.len }
	}
}

pub fn (d &DataBuilder) build() []u8 {
	mut b := []u8{len: d.len}
	d.func(mut b)
	binary.big_endian_put_u16(mut b, u16(d.len - 2))
	unsafe {
		d.func = nil
		d.len = 2
	}
	return b
}

fn new_data_builder() DataBuilder {
	return DataBuilder{
		func: unsafe { nil }
		len: 2
	}
}

fn main() {
	mut data := new_data_builder()
	data.add(1)
	data.add(2)
	data.add(3)
	data.add(4)
	data.add(5)
	data.add('1')
	data.add('22')
	data.add('334455')
	data.add(rand.string(rand.int_in_range(10, 100)!))
	data.add('5')
	data.add(1)
	data.add(2)
	data.add(3)
	data.add(4)
	data.add(5)
	ret := data.build()
	dump(binary.big_endian_u16_at(ret, 0) == ret[2..].len)
	dump(ret)
}

```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.0 ad1d5e7.b06811c

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.0 ad1d5e7.b06811c
OS: windows, Microsoft Windows Server 2022 Datacenter v25379 64 λ
Processor: 8 cpus, 64bit, little endian, 

getwd: D:\vproj\test
vexe: D:\Tools\Compiler\v\v.exe
vexe mtime: 2023-07-14 21:39:06

vroot: OK, value: D:\Tools\Compiler\v
VMODULES: OK, value: C:\Users\Administrator\.vmodules
VTMP: OK, value: C:\Users\Administrator\AppData\Local\Temp\v_0

Git version: git version 2.40.1.windows.1
Git vroot status: weekly.2023.28-27-gb06811cb-dirty
.git/config present: true

CC version: Error: exec failed (CreateProcess) with code 2: 系统找不到指定的文件。

 cmd: cc --version
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3650,6 +3650,7 @@ fn (mut c Checker) smartcast(mut expr ast.Expr, cur_type ast.Type, to_type_ ast.
 			mut smartcasts := []ast.Type{}
 			mut is_already_casted := false
 			mut orig_type := 0
+			mut is_inherited := false
 			if mut expr.obj is ast.Var {
 				is_mut = expr.obj.is_mut
 				smartcasts << expr.obj.smartcasts
@@ -3657,6 +3658,7 @@ fn (mut c Checker) smartcast(mut expr ast.Expr, cur_type ast.Type, to_type_ ast.
 				if orig_type == 0 {
 					orig_type = expr.obj.typ
 				}
+				is_inherited = expr.obj.is_inherited
 			}
 			// smartcast either if the value is immutable or if the mut argument is explicitly given
 			if (!is_mut || expr.is_mut) && !is_already_casted {
@@ -3667,6 +3669,7 @@ fn (mut c Checker) smartcast(mut expr ast.Expr, cur_type ast.Type, to_type_ ast.
 					pos: expr.pos
 					is_used: true
 					is_mut: expr.is_mut
+					is_inherited: is_inherited
 					smartcasts: smartcasts
 					orig_type: orig_type
 				})

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4348,6 +4348,9 @@ fn (mut g Gen) ident(node ast.Ident) {
 						} else {
 							mut is_ptr := false
 							if i == 0 {
+								if node.obj.is_inherited {
+									g.write(closure_ctx + '->')
+								}
 								g.write(name)
 								if node.obj.orig_type.is_ptr() {
 									is_ptr = true

--- a/vlib/v/tests/closure_with_sumtype_var_test.v
+++ b/vlib/v/tests/closure_with_sumtype_var_test.v
@@ -1,0 +1,75 @@
+import encoding.binary
+import rand
+
+type DataType = int | string
+
+struct DataBuilder {
+mut:
+	func fn (mut []u8)
+	len  int
+}
+
+pub fn (mut d DataBuilder) add(v DataType) {
+	func := d.func
+	len := d.len
+	d.func = fn [v, len, func] (mut b []u8) {
+		if !isnil(func) {
+			func(mut b)
+		}
+		match v {
+			int {
+				binary.big_endian_put_u32_at(mut b, u32(v), len)
+			}
+			string {
+				binary.big_endian_put_u16_at(mut b, u16(v.len), len)
+				for i := 0; i < v.len; i++ {
+					b[len + 2 + i] = v[i]
+				}
+			}
+		}
+	}
+	d.len += match v {
+		int { 4 }
+		string { 2 + v.len }
+	}
+}
+
+pub fn (d &DataBuilder) build() []u8 {
+	mut b := []u8{len: d.len}
+	d.func(mut b)
+	binary.big_endian_put_u16(mut b, u16(d.len - 2))
+	unsafe {
+		d.func = nil
+		d.len = 2
+	}
+	return b
+}
+
+fn new_data_builder() DataBuilder {
+	return DataBuilder{
+		func: unsafe { nil }
+		len: 2
+	}
+}
+
+fn test_closure_with_sumtype_var() {
+	mut data := new_data_builder()
+	data.add(1)
+	data.add(2)
+	data.add(3)
+	data.add(4)
+	data.add(5)
+	data.add('1')
+	data.add('22')
+	data.add('334455')
+	data.add(rand.string(rand.int_in_range(10, 100)!))
+	data.add('5')
+	data.add(1)
+	data.add(2)
+	data.add(3)
+	data.add(4)
+	data.add(5)
+	ret := data.build()
+	dump(ret)
+	assert binary.big_endian_u16_at(ret, 0) == ret[2..].len
+}

